tool/tsh/access_request_test.go

```diff
@@ -0,0 +1,149 @@
+)
+
+func newKeygen(precompute int) *native.Keygen {
+	fakeClock := clockwork.NewFakeClockAt(time.Date(2016, 9, 8, 7, 6, 5, 0, time.UTC))
+
+	return native.New(
+		context.TODO(),
+		native.PrecomputeKeys(precompute),
+		native.SetClock(fakeClock))
+
+}
+
+func newKey(keygen *native.Keygen, roles []string) (*client.Key, error) {
+	priv, pub, err := keygen.GenerateKeyPair("")
+	if err != nil {
+		return nil, err
+	}
+	cert, err := keygen.GenerateUserCertWithoutValidation(
+		services.UserCertParams{
+			PrivateCASigningKey:   priv,
+			CASigningAlg:          defaults.CASignatureAlgorithm,
+			PublicUserKey:         pub,
+			Username:              "some-user",
+			AllowedLogins:         []string{},
+			TTL:                   time.Hour,
+			Roles:                 roles,
+			CertificateFormat:     teleport.CertificateFormatStandard,
+			PermitAgentForwarding: true,
+			PermitPortForwarding:  true,
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	key := &client.Key{
+		Priv: priv,
+		Pub:  pub,
+		Cert: cert,
+	}
+
+	return key, nil
+}
+
+func TestPassingNilToIsNotFoundReturnsFalse(t *testing.T) {
+	require.False(t, trace.IsNotFound(nil))
+}
+
+func TestKeysTriggerRoleChange(t *testing.T) {
+	// We can run this top-level test in parallel with the other parallel tests,
+	// but given that all the sub-tests interact with the same Keygen, they
+	// should probably run in series within this parent.
+	t.Parallel()
+
+	testCases := []struct {
+		desc         string
+		rolesOld     []string
+		rolesNew     []string
+		expectChange bool
+	}{
+		{
+			desc:         "Same",
+			rolesOld:     []string{"alpha", "beta", "gamma"},
+			rolesNew:     []string{"alpha", "beta", "gamma"},
+			expectChange: false,
+		}, {
+			desc:         "Out of order",
+			rolesOld:     []string{"beta", "gamma", "alpha"},
+			rolesNew:     []string{"alpha", "beta", "gamma"},
+			expectChange: false,
+		}, {
+			desc:         "Empty",
+			rolesOld:     []string{},
+			rolesNew:     []string{},
+			expectChange: false,
+		}, {
+			desc:         "Removal",
+			rolesOld:     []string{"alpha", "beta", "gamma", "delta"},
+			rolesNew:     []string{"alpha", "beta", "gamma"},
+			expectChange: true,
+		}, {
+			desc:         "Addition",
+			rolesOld:     []string{"alpha", "beta", "gamma"},
+			rolesNew:     []string{"alpha", "beta", "gamma", "delta"},
+			expectChange: true,
+		}, {
+			desc:         "Total Replacement",
+			rolesNew:     []string{"alpha", "beta", "gamma"},
+			rolesOld:     []string{"delta", "epsilon", "zeta"},
+			expectChange: true,
+		}, {
+			desc:         "Duplicate replacement",
+			rolesOld:     []string{"alpha", "alpha", "gamma"},
+			rolesNew:     []string{"alpha", "beta", "gamma"},
+			expectChange: true,
+		}, {
+			desc:         "Duplicate insertion",
+			rolesOld:     []string{"alpha", "beta", "gamma"},
+			rolesNew:     []string{"alpha", "alpha", "gamma"},
+			expectChange: true,
+		},
+	}
+	keyGen := newKeygen(len(testCases))
+	defer keyGen.Close()
```

Comment:
Prefer `t.Cleanup` to `defer` in tests.

Reply:
Happy to oblige, but what's the rationale?